test/clonevolume_tester.py

```diff
@@ -0,0 +1,262 @@
+        mock_3parclient.getCPG.return_value = {}
+
+
+# Online copy with flash cache - add to vvset fails
+class TestCloneWithFlashCacheAddVVSetFails(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp,
+                                    {u"Err": u'Not found (HTTP 404) - fake'})
+
+        # Check required WSAPI calls were made
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.createVolumeSet.assert_called()
+        mock_3parclient.addVolumeToVolumeSet.assert_called()
+        mock_3parclient.deleteVolumeSet.assert_called()
+
+        # TODO: This is not happening at the moment and would make
+        # the unit test fail
+        mock_3parclient.deleteVolume.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume_flash_cache
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        # Make addVolumeToVolumeSet fail by throwing exception
+        mock_3parclient.addVolumeToVolumeSet.side_effect = [exceptions.HTTPNotFound('fake')]
+
+
+# CHAP enabled makes Offline copy flow to execute
+class TestCloneWithCHAP(CloneVolumeUnitTest):
+    def override_configuration(self, config):
+        config.hpe3par_iscsi_chap_enabled = True
+
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.getVolumeMetaData.assert_called()
+        mock_3parclient.createVolume.assert_called()
+        mock_3parclient.copyVolume.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        mock_3parclient.getVolumeMetaData.return_value = {'value': True}
+        mock_3parclient.getTask.return_value = {'status': data.TASK_DONE}
+
+# TODO: This is already covered in other tests above
+class TestCloneWithoutCHAP(CloneVolumeUnitTest):
+    pass
+
+
+# Override WS-API-Version to have lower value
+class TestCloneUnsupportedDedupVersion(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
```

Comment:
This looks to be an invalid test case for me
Cloning a deduped volume on an unsupported array is not possible. Sorce volume will not be available to make a clone in this case(If wsapi version is not supported for creating dedup volume , the copy operation will fail as the user do not have source volume in etcd available)

Reply:
Okay. In which case even the main flow should not have this check.